pkg/compactor/block_upload.go

```diff
@@ -247,50 +278,97 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 
 	dst := path.Join(blockID.String(), pth)
 
-	level.Debug(logger).Log("msg", "uploading block file to bucket", "destination", dst, "size", r.ContentLength)
+	level.Debug(blockLogger).Log("msg", "uploading block file to bucket", "destination", dst, "size", r.ContentLength)
 	reader := bodyReader{r: r}
 	if err := userBkt.Upload(ctx, dst, reader); err != nil {
-		level.Error(logger).Log("msg", "failed uploading block file to bucket", "operation", op, "destination", dst, "err", err)
+		level.Error(blockLogger).Log("msg", "failed uploading block file to bucket", "operation", op, "destination", dst, "err", err)
 		// We don't know what caused the error; it could be the client's fault (e.g. killed
 		// connection), but internal server error is the safe choice here.
 		http.Error(w, "internal server error", http.StatusInternalServerError)
 		return
 	}
 
-	level.Debug(logger).Log("msg", "finished uploading block file to bucket", "path", pth)
+	level.Debug(blockLogger).Log("msg", "finished uploading block file to bucket", "path", pth)
 
 	w.WriteHeader(http.StatusOK)
 }
 
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) error {
-	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+func (c *MultitenantCompactor) validateAndCompleteBlockUpload(blockLogger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) {
+	level.Debug(blockLogger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+
+	var ctx context.Context
+	var cancel context.CancelFunc
+
+	if c.compactorCfg.BlockUpload.ValidationHeartbeatTimeout > 0 {
```

Comment:
I'll brain dump a bit here.

This use of the heartbeat timeout is incorrect, since it confuses timing out based off the last heartbeat time with timing out based off the duration of the entire validation. I'm considering adding the latter as a separate option, since currently a validation can take an indefinite amount of time as long as it continues to heartbeat.

The heartbeating is also currently unaware of when a validation has been restarted, allowing for concurrent validations for the same block to heartbeat over one another. Adding reads to the heartbeat cycle and maybe a validation ULID could be added to double check for this.

The heartbeats also currently prevent partial block cleanup from occurring. Maybe the cleanup can be smarter when interacting with block uploads by recognizing the renamed upload meta file.

Reply:
One other idea I find interesting is to abandon the idea of verification initiating from a finish handler, and instead have the verification be picked up as if it were any other normal compaction. The feedback mechanism of determining why an upload failed gets even worse in that scenario, but it sidesteps the issue of creating a new path to load balance. Maybe a remedy to this would be to have the verification process be executable from mimirtool locally so it would derive the same error encountered on the server. Two potential complications to that would be mismatching versions or configuration modifying the verification process.

Reply:
> The heartbeating is also currently unaware of when a validation has been restarted, allowing for concurrent validations for the same block to heartbeat over one another.

We can't start new validation if there is already ongoing validation. Only if validation file wasn't updated for some time (and thus we assume that compactor has restarted), then we allow new validation.

> The heartbeats also currently prevent partial block cleanup from occurring.

I'm not sure I follow. Heartbeats are only happening during validation.

> Maybe a remedy to this would be to have the verification process be executable from mimirtool locally so it would derive the same error encountered on the server.

Having an option to run validation locally by using `mimirtool` was a planned feature.

Reply:
It seems my idea wasn't popular for the sake of time, but I'll respond for technical discussion's sake (no hard feelings!)

> We can't start new validation if there is already ongoing validation. Only if validation file wasn't updated for some time (and thus we assume that compactor has restarted), then we allow new validation.

The detection of whether there is an ongoing validation depends on the read of the validation file, which does not stop concurrent callers.

The core problem is using marker files on object storage for immediate coordination when they are better suited for eventual coordination, otherwise they allow for time of check vs time of use issues. That's the crux of why I proposed a more eventual model.

> I'm not sure I follow. Heartbeats are only happening during validation. And we don't want partial block cleanup to happen while validation is still going on.

I found the interaction surprising. Was this intentional? We should probably document it if so, since I wasn't involved when this was originally written.

I agree you don't want partial cleanup to happen during verification. It can though. Imagine the case of not calling finish upload right away.

The partial cleanup (as far as I understand it?) was written for ingesters shipping blocks, which is similar to what is being done here except with a user instead of an ingester on the other end. `checkBlockState` could check for the delete marker of a block to tell you verifying isn't worth it if you managed to wait that long. It's not a guarantee, but could help.

> Having an option to run validation locally by using mimirtool was a planned feature.

Nice!